code/crates/sync/src/scoring.rs

```diff
@@ -0,0 +1,731 @@
+    pub fn new(strategy: impl ScoringStrategy + 'static) -> Self {
+        Self {
+            scores: HashMap::new(),
+            strategy: Box::new(strategy),
+        }
+    }
+
+    /// Update a peer's score based on the result of a sync request, recording the result in metrics.
+    /// Returns the new score.
+    pub fn update_score_with_metrics(
+        &mut self,
+        peer_id: PeerId,
+        result: SyncResult,
+        metrics: &Metrics,
+    ) -> Score {
+        let new_score = self.update_score(peer_id, result);
+        metrics.observe_score(peer_id, new_score);
+        new_score
+    }
+
+    /// Update a peer's score based on the result of a sync request.
+    /// Returns the new score.
+    pub fn update_score(&mut self, peer_id: PeerId, result: SyncResult) -> Score {
+        let peer_score = self
+            .scores
+            .entry(peer_id)
+            .or_insert_with(|| PeerScore::new(self.strategy.initial_score(peer_id)));
+
+        let previous_score = peer_score.score;
+
+        debug!("Updating score for peer {peer_id}");
+        debug!("  Result = {result:?}");
+        debug!("    Prev = {previous_score}");
+
+        let new_score = self.strategy.update_score(previous_score, result);
+        debug!("     New = {new_score}");
+
+        peer_score.score = new_score;
+        peer_score.last_update = Instant::now();
+
+        new_score
+    }
+
+    /// Get the current score for a peer
+    pub fn get_score(&self, peer_id: &PeerId) -> Score {
+        self.scores
+            .get(peer_id)
+            .map(|p| p.score)
+            .unwrap_or(self.strategy.initial_score(*peer_id))
+    }
+
+    /// Get all peer scores
+    pub fn get_scores(&self) -> &HashMap<PeerId, PeerScore> {
+        &self.scores
+    }
+
+    /// Select a peer using weighted probabilistic selection
+    pub fn select_peer<R: Rng>(&self, peers: &[PeerId], rng: &mut R) -> Option<PeerId> {
+        if peers.is_empty() {
+            return None;
+        }
+
+        let scores = peers.iter().map(|id| self.get_score(id).max(0.0));
+
+        // Sample from peers using a weighted distribution based on their scores
+        let distr = WeightedIndex::new(scores).ok()?;
+        let index = distr.sample(rng);
+
+        assert!(index < peers.len(), "Index out of bounds");
+        Some(peers[index])
+    }
+
+    /// Prune peers whose scores have not been updated for the specified duration,
+    /// effectively resetting their score to the initial score.
+    ///
+    /// A peer might be inactive because they were not picked for a long time
+    /// due to their score being very low. Resetting their score gives them a chance to participate again.
+    ///
+    /// Note that by resetting the score we can also reduce the score of a peer,
+    /// if the peer had a high score but was inactive for a long time.
+    pub fn prune_inactive_peers(&mut self, inactive_threshold: Duration) {
```

Comment:
The word prune makes me think of cutting the network connection. How about `reset_inactive_peers`?